### PR TITLE
Statistics: Implement statistic variables

### PR DESCRIPTION
**Pull request by bstoeger**
### Describe the pull request:

- [ ] Bug fix
- [ ] Functional change
- [x] New feature
- [ ] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:

The first in a number of PRs for reviewing the statistics module. This implements the concept of "dive variable" that can be plotted as dependent or independent variable.

Yes, the code is template heavy and at time convoluted. It looks worse than it is, because it doesn't use any TMP (template meta programming) only old-fashioned templates in the sense of type-checked macros. I'm not very proud of it and I'm sure it can be made more space efficient with runtime-polymorphism. However, it allowed me to rather quickly implement this, without too much planning. It has proven to be quite easily expandable.

If any of the constructs are unclear (I suspect there are a few of them), feel free to ask.

### Changes made:

1) A few preparatory commits (the SVG thing is not necessary, but let's bring it in now).
2) Add the statsvariable.cpp source file and compile it.

### Mentions:

@dirkhh

### CODE REVIEW

File: core/imagedownloader.cpp
Comment (dirkhh) [inline core/imagedownloader.cpp]: I wonder if it wouldn't be more consistent to move this function into qthelper as well - those two seem rather closely related, having one here, one there seems... odd

File: core/gas.c
Comment (dirkhh) [inline core/gas.c]: So I am confused by that. 
The definition of **normoxic** that I am familiar with is a Trimix gas that has roughly normal oxygen content. Every agency thinks about the details differently, but a quick google search landed me here http://www.ddcbuildsite.com/index.php/easyblog/entry/normoxic-trimix-diving-diving-deep-with-a-clearer-head which suggests O2 values between 18-21% -- I was used to 19-22%
I have never heard of the term **normoxic** (which, etymologically seems to imply something like 'normal oxygen') to be triggered by the H2 percentage as in this formula...
If we want to go that route, then one other term that I hear thrown around a lot with trimix divers (full disclosure, I'm a TDI tech dive master and have dive-mastered (shouldn't say 'taught') advanced trimix classes in the past) is **hypoxic** (so hypo or below oxygen) for trimix gasses with an O2 content below whatever you pick for the lower threshold above - in my training 18/40 was considered **hypoxic**, but 19/40 was considered **normoxic**

I am NOT saying that these are the definitions we have to use - all I'm saying is that the definition above has me rather confused...

File: core/gas.h
Comment (dirkhh) [inline core/gas.h:32]: I realize that this is missing another significant gas. **Oxygen** (O2 >= 98%)

File: stats/statsvariables.cpp
Comment (dirkhh) [inline stats/statsvariables.cpp:119]: so three commits ago you added the gas types. There you didn't have **oxygen**, but you had **normoxic** (and were IMHO also missing **hypoxic**)
Maybe that's a different use case, but that would be at least worth a comment, right?

File: stats/statsvariables.cpp
Comment (dirkhh) [inline stats/statsvariables.cpp:138]: not sure if this is permanently prevented somehow, but `o2+he > 1000` or either of the individual values `< 0` or `> 1000` all seem like valid `invalid` criteria

File: stats/statsvariables.cpp
Comment (dirkhh) [inline stats/statsvariables.cpp:174]: I think it's fine to implement this in two places, but this should mention 
https://github.com/subsurface/subsurface/blob/edd19d4725153e6428e7b2e1b6594b4168601a39/core/divelist.c#L32-L39

File: stats/statsvariables.cpp
Comment (dirkhh) [inline stats/statsvariables.cpp]: I know, bike shedding, but wouldn't it be much easier to just do
```
if (!is_invalid_value(v))
        vec.push_back({ v, d });
```

File: stats/statsvariables.cpp
Comment (dirkhh) [inline stats/statsvariables.cpp]: premature

File: stats/statsvariables.cpp
Comment (dirkhh) [inline stats/statsvariables.cpp]: premature

File: stats/statsvariables.cpp
Comment (dirkhh) [inline stats/statsvariables.cpp:817]: stupid bike shedding comment... leap seconds...?

File: stats/statsvariables.cpp
Comment (dirkhh) [inline stats/statsvariables.cpp:950]: I guess this uses the preference, not a variable... but it seemed that I was never given an option to use anything but what's used in the preferences, so...
https://github.com/subsurface/subsurface/blob/edd19d4725153e6428e7b2e1b6594b4168601a39/core/units.h#L27

File: stats/statsvariables.cpp
Comment (dirkhh) [inline stats/statsvariables.cpp]: that's... surprising... but I guess it qualifies as "good enough"

File: stats/statsvariables.cpp
Comment (dirkhh) [inline stats/statsvariables.cpp]: we are binning weight in meters or feet? I doubt that

File: stats/statsvariables.cpp
Comment (dirkhh) [inline stats/statsvariables.cpp]: I always thought this was an odd choice (and it's been around forever)

File: stats/statsvariables.cpp
Comment (dirkhh) [inline stats/statsvariables.cpp:1338]: that gas binner is... crazy... when testing the code I kept trying to figure out how it bins my gases. Just wow.

**Review by dirkhh — COMMENTED** [review]
That is one monster commit at the end. But it makes sense to just get it over with.
It's making my brain melt and while I did add quite a few comments, I can't claim to have reviewed the correctness of every single line. It uses quite a bit of syntax constructs that I do understand, but that aren't comfortable, so I keep reminding myself what they actually mean. Which makes review harder.

There are a couple of issues that I mention in comments that stop me from approving.

File: core/gas.c
Comment (bstoeger) [reply core/gas.c]: I originally had hypoxic, but I believe that @willemferguson objected.
See https://github.com/subsurface/subsurface/pull/3052#issuecomment-731947533 and follow-up discussion.
Perhaps I misunderstood things. The good thing is that this is trivial to change by changing gas.h and gas.c.

File: stats/statsvariables.cpp
Comment (bstoeger) [reply stats/statsvariables.cpp:119]: Yes, the former was for binning with a "general type". This is for formatting bins that use the actual composition of the gases. Which is fun, because it means that we have two-dimensional bins. I shortly entertained the idea, but then refrained from doing 3D histograms (O2 vs H2 vs Count). 😀

File: stats/statsvariables.cpp
Comment (bstoeger) [reply stats/statsvariables.cpp:817]: I'm not qualified enough to make a comment on that. I.e. I don't know how leap seconds are treated by the C library.

In any case, the deviation from linearity should be negligible.

File: stats/statsvariables.cpp
Comment (bstoeger) [reply stats/statsvariables.cpp:138]: Yes - we never bin invalid gases, so invalid gas-bins will never exist:
```
                        struct gasmix mix = d->cylinders.cylinders[i].gasmix;
                        if (gasmix_is_invalid(mix))
                                continue;
```
Added a comment.

File: stats/statsvariables.cpp
Comment (bstoeger) [reply stats/statsvariables.cpp:174]: These are related, but not the same: `get_dive_gas` gets the "maximum" gas. The `operator<` is used to establish a total order of gas bins.

File: stats/statsvariables.cpp
Comment (bstoeger) [reply stats/statsvariables.cpp]: done.

File: stats/statsvariables.cpp
Comment (bstoeger) [reply stats/statsvariables.cpp]: done.

File: stats/statsvariables.cpp
Comment (bstoeger) [reply stats/statsvariables.cpp]: done.

File: stats/statsvariables.cpp
Comment (bstoeger) [reply stats/statsvariables.cpp:950]: This macro does something completely different, no? It is used of you want to provide two "nice" numbers.

In any case - the code was very inconsistent : Sometimes it would respect the "metric" variable, sometimes use the preferences flag. After some thinking I concluded that the correct thing to do is using the variable. For now, this is an obscure difference. But what if we save the statistics into presets similar to the filters? The user may end up with loading a binner for a different unit system than in the preferences. Therefore the binners now all format according to their proper type, just to be sure. The numeric representations don't have that problem and therefore use the preferences flag.

File: stats/statsvariables.cpp
Comment (bstoeger) [reply stats/statsvariables.cpp]: The unit is now made explicit.

File: stats/statsvariables.cpp
Comment (bstoeger) [reply stats/statsvariables.cpp]: done.

File: stats/statsvariables.cpp
Comment (bstoeger) [reply stats/statsvariables.cpp]: I think it kind of makes sense, because standard leads come in half-pound and full-kg units, no?

File: stats/statsvariables.cpp
Comment (bstoeger) [reply stats/statsvariables.cpp:1338]: It is a bit brutal, indeed. There are two dimensions to this:
- You can bin coarsely by type and finely by percentage.
- The gas is selected depending on the variable (max o2, max he). Not that it is too useful, but maybe someone wants to plot bottom gas vs. deco gas?

File: core/imagedownloader.cpp
Comment (bstoeger) [reply core/imagedownloader.cpp]: done. That said, qthelper.cpp really needs to be split up.

File: core/gas.h
Comment (bstoeger) [reply core/gas.h:32]: I added oxygen, but would suggest that you come to a conclusion with respect to hypoxic with @willemferguson.

**Comment by bstoeger** [comment]
PS: added a weekday-binner, as that appeared to be low-hanging fruit.

**Comment by bstoeger** [comment]
It is not really related, but since it is so trivial I added the colorscheme commit, so that we do some progress.

Waiting for @dirkhh's review, I will try to improve the highlighting of the scatter plots.

**Comment by bstoeger** [comment]
Added the legend, since it is also small and self-contained.

@dirkhh: if this is too fast and you'd prefer to merge things in small pieces, let me know.

**Comment by bstoeger** [comment]
Also let me know if you would prefer the legend box to have rounded corners, a different color, etc. These are all things that are trivial to implement.

**Comment by dirkhh** [comment]
Everything needs to have rounded corners, right? That looks almost always better.
I just got up and as my family is still asleep, I'll try to get as far as I can with this. I'll look through your responses and the new commits right now.

File: core/gas.c
Comment (dirkhh) [reply core/gas.c]: I can look through that discussion again and see how we got here - I'll focus on the rest, first, to optimize time :-)

File: stats/statsvariables.cpp
Comment (dirkhh) [reply stats/statsvariables.cpp:817]: It was meant as a joke among all the others... mainly because you divided by `86400.0` which of course was intended to force floating point math, but to my silly eyes looked like a statement of precision...

File: stats/statsvariables.cpp
Comment (dirkhh) [reply stats/statsvariables.cpp:950]: Yes, that's a good point. The ability to have bins in a different unit system than the one currently set in preferences might avoid pain later.

File: stats/statsvariables.cpp
Comment (dirkhh) [reply stats/statsvariables.cpp]: not really. leads almost never weigh what people claim they do... and they are typically multiples of US pounds or half KG. So 10 lbs are 4.5kg. And because Linus and I both 'thought' metric a lot, but dove mostly with US weights, we wanted it to look correct both ways. Which is of course dumb, because if you dive with 3kg in Egypt then that's 6.6lbs :-)

File: stats/statsvariables.cpp
Comment (dirkhh) [reply stats/statsvariables.cpp:1338]: There are so many options there. I'm not convinced I fully understood everything you offer.

File: stats/statsvariables.cpp
Comment (dirkhh) [reply stats/statsvariables.cpp:174]: Right - I was wondering if you want to comment that we use a similar order elsewhere and refer to it. Simply for consistency. I did get that these are different functions :-)

File: core/gas.c
Comment (dirkhh) [reply core/gas.c]: Ok, found it: https://github.com/subsurface/subsurface/pull/3052#issuecomment-732253281
> With respect to the hypoxic gas class, I would recommend that you merge it with trimix. Full trimix dives are characterised by a hypoxic oxygen concentration (< 18%) and Helium < 30%. In contrast, normoxic dives are usually defined by oxygen around 20% and helium <= 30%. I would say that in a grey area the critical distinguishing factor is the oxygen concentration.

So this is very much what I talked about above, just with `< 18` instead of `<= 18` as in my training.
But I think his underlying comment kinda was to drop hypoxic (and not to add normoxic) and just have Air, Nitrox, Trimix (and maybe Oxygen as extreme Nitrox).
That said, I think it would be interesting to list all the classes (we can always add 12 million configuration options later... on second thought, let's NOT add all those options).
Outside of training dives I don't think you'll find a lot of trimix with O2 above 22%, so we could call that "Trimix" and the other two "Normoxic Trimix" and "Hypoxic Trimix". Makes sense?

File: stats/legend.cpp
Comment (dirkhh) [inline stats/legend.cpp:49]: I realized later that this was a bad choice in dark mode. But we can fix that once everything is merged, I don't want to hold this up here.

File: stats/legend.cpp
Comment (dirkhh) [inline stats/legend.cpp:69]: that does seem like a LOT. But you are right, easy to fine tune later.

**Review by dirkhh — APPROVED** [review]
I think the minor questions left are easily resolved.
Please consider the comments that I made about the gases, but once you decided on what to do (just Trimix, or three flavors of Trimix) and fixed the determination to be purely based on the O2 content, this is good to go

File: core/gas.c
Comment (bstoeger) [reply core/gas.c]: > Outside of training dives I don't think you'll find a lot of trimix with O2 above 22%, so we could call that "Trimix" and the other two "Normoxic Trimix" and "Hypoxic Trimix". Makes sense?

Fine with me. Note that I think that @willemferguson's argument was that, precisely because hypoxic is the standard Trimix, mixtures <= 18% should be called "Trimix" without further qualifier. I will just qualify all three and you can then discuss which one is the "Trimix" version. FWIW, according to Wikipedia hyperoxic Trimix is also called "TriOx".

File: stats/statsvariables.cpp
Comment (bstoeger) [reply stats/statsvariables.cpp]: Oh, I just realize that it is exactly the other way around than I would have expected. Should this condition be inverted?

File: stats/statsvariables.cpp
Comment (bstoeger) [reply stats/statsvariables.cpp:174]: Comment added.

File: stats/statsvariables.cpp
Comment (dirkhh) [reply stats/statsvariables.cpp]: What you have here is what we have elsewhere. One decimal for metric, no decimal for pounds. It's inconsistent and stupid. We should lead the way and simply always offer one decimal :-)

Comment (bstoeger) [reply stats/statsvariables.cpp]: I inverted the condition: it is now one decimal for kg and no decimals for lbs. Of course we might simply change to one decimal fraction for all.

File: core/gas.c
Comment (dirkhh) [inline core/gas.c:154]: I like it.
I'm not sure if there are people diving with hypoxic nitrox, TBH - you call that air here (and I hadn't mentioned it earlier). But again, I think this is fine.

**Comment by bstoeger** [comment]
> Everything needs to have rounded corners, right? That looks almost always better.

Ok, let's do that later. I just realized that this means overwriting the `paint()` function, which I don't feel like doing currently. See `ToolTipItem::paint()`.

**Review by dirkhh — APPROVED** [review]
In order to keep moving, I think we should merge this once the CI passes...

File: stats/statsvariables.cpp
Comment (bstoeger) [reply stats/statsvariables.cpp]: > What you have here is what we have elsewhere. One decimal for metric, no decimal for pounds. 

Which is funny, because I'm quite sure that this was just a mix-up from my side. Replaced by `return 1;`.

**Comment by bstoeger** [comment]
> In order to keep moving, I think we should merge this once the CI passes...

Fine - will merge. There will surely be issues popping up, but we can fix them later.

**Pull request merged**
